table/table_properties.cc

```diff
@@ -282,24 +307,44 @@ const std::string TablePropertiesNames::kFastCompressionEstimatedDataSize =
 #ifndef NDEBUG
 void TEST_SetRandomTableProperties(TableProperties* props) {
   Random* r = Random::GetTLSInstance();
-  // For now, TableProperties is composed of a number of uint64_t followed by
-  // a number of std::string, followed by some extras starting with
-  // user_collected_properties.
-  uint64_t* pu = &props->orig_file_number;
-  assert(static_cast<void*>(pu) == static_cast<void*>(props));
-  std::string* ps = &props->db_id;
-  const uint64_t* const pu_end = reinterpret_cast<const uint64_t*>(ps);
-  const std::string* const ps_end =
-      reinterpret_cast<const std::string*>(&props->user_collected_properties);
 
+  std::pair<const uint64_t*, const uint64_t*> uint64t_prop_start_end_pos =
+      TEST_GetUint64TPropStartEndPosition(props);
+  uint64_t* pu = const_cast<uint64_t*>(uint64t_prop_start_end_pos.first);
+  const uint64_t* const pu_end = uint64t_prop_start_end_pos.second;
   for (; pu < pu_end; ++pu) {
     *pu = r->Next64();
   }
-  assert(static_cast<void*>(pu) == static_cast<void*>(ps));
+
+  std::pair<const std::string*, const std::string*> string_prop_start_end_pos =
+      TEST_GetStringPropStartEndPosition(props);
+  std::string* ps = const_cast<std::string*>(string_prop_start_end_pos.first);
+  const std::string* const ps_end = string_prop_start_end_pos.second;
   for (; ps < ps_end; ++ps) {
     *ps = r->RandomBinaryString(13);
   }
 }
 #endif
 
+std::pair<const uint64_t*, const uint64_t*> TEST_GetUint64TPropStartEndPosition(
+    const TableProperties* const_props) {
+  TableProperties* props = const_cast<TableProperties*>(const_props);
+  uint64_t* pu = &props->orig_file_number;
+  assert(static_cast<void*>(pu) == static_cast<void*>(props));
+  const uint64_t* pu_end = reinterpret_cast<const uint64_t*>(&props->db_id);
+  return std::pair<const uint64_t*, const uint64_t*>(pu, pu_end);
+}
+
+std::pair<const std::string*, const std::string*>
+TEST_GetStringPropStartEndPosition(const TableProperties* const_props) {
+  TableProperties* props = const_cast<TableProperties*>(const_props);
+  std::string* ps = &props->db_id;
+  assert(static_cast<void*>(const_cast<uint64_t*>(
+             TEST_GetUint64TPropStartEndPosition(props).second)) ==
+         static_cast<void*>(ps));
+  const std::string* ps_end =
+      reinterpret_cast<const std::string*>(&props->user_collected_properties);
+  return std::pair<const std::string*, const std::string*>(ps, ps_end);
+}
```

Comment:
This is not easy on the eyes. Do we need it? I thought we were adding the sizes of the strings or something similarly understandable before.

Reply:
OK it makes enough sense to me. Just make sure to eliminate all `TEST_` in non-test code, `const_cast`s, and `reinterpret_cast`s.

Reply:
This is suggested by https://github.com/facebook/rocksdb/pull/9748#discussion_r841971925. Added a [test](https://github.com/facebook/rocksdb/pull/9748/files#diff-d17c3d6f851ee716cc49892ea8712195b0aa032285c654370d29b202d07e4b0bR414) /[comment](https://github.com/facebook/rocksdb/pull/9748/files#diff-5653feed9ef035ce3869aa3c8c2cae720229948bf901638888c05656f7219f08R15) for this too.

Although I still agree it's not easy to understand nor encouraged. I was hoping to name it TEST_XXX to discourage usage of it but I will remove the "TEST_" as suggested.

Reply:
> Although I still agree it's not easy to understand nor encouraged. I was hoping to name it TEST_XXX to discourage usage of it

I think if the C++ is not safe we should actually only use it in tests, not here either. But let's see if there's a proof that `reinterpret_cast<const std::string*>(&props->user_collected_properties)` is guaranteed to be the right string pointer for us to compare against.

Reply:
> What if `user_collected_properties` has different alignment from string?

I believe we aren't reading any value pointed by "reinterpret_cast<const std::string*>(&props->user_collected_properties)" cuz the value won't be the correct as it's reinterpreted as string "incorrectly". We only use the address for purpose of advancing the pointer and the "reinterpret_cast" is only for `ps < ps_end` in the for statement. 

cc @pdillinger as it seems that you started this [hack](https://github.com/facebook/rocksdb/blame/36bc3da97fcb6aa2df6e43df27552f49b762d61e/table/table_properties.cc#L283) and correct me if I'm wrong.

Reply:
> OK it makes enough sense to me. Just make sure to eliminate all TEST_ in non-test code, const_casts, and reinterpret_casts (ideally).

Will do.

Reply:
Done by (a) checking non TEST_ call-sites of GetXXXPropStartEndPosition do not have const_casts, and reinterpret_casts (b) putting all the const_casts in GetXXXPropStartEndPosition in assertion code (c) can't do anything with `reinterpret_cast ` within the GetXXXPropStartEndPosition though

Reply:
> I think if the C++ is not safe we should actually only use it in tests, not here either. But let's see if there's a proof that `reinterpret_cast<const std::string*>(&props->user_collected_properties)` is guaranteed to be the right string pointer for us to compare against.

Is something like this https://github.com/facebook/rocksdb/pull/9748/files#diff-d17c3d6f851ee716cc49892ea8712195b0aa032285c654370d29b202d07e4b0bR434-R435 efficient? Casting both to void* and check address equality

Reply:
> I believe we aren't reading any value pointed by "reinterpret_cast<const std::string*>(&props->user_collected_properties)" cuz the value won't be the correct as it's reinterpreted as string "incorrectly". We only use the address for purpose of advancing the pointer and the "reinterpret_cast" is only for ps < ps_end in the for statement.

Would we dereference an invalid pointer if there were any padding bytes between compression_options and user_collected_properties?

Reply:
For example add:

```
$ git diff include/rocksdb/table_properties.h 
diff --git a/include/rocksdb/table_properties.h b/include/rocksdb/table_properties.h
index 0c91fa854..fa157f7ee 100644
--- a/include/rocksdb/table_properties.h
+++ b/include/rocksdb/table_properties.h
@@ -285,7 +285,7 @@ struct TableProperties {
   std::string compression_options;
 
   // user collected properties
-  UserCollectedProperties user_collected_properties;
+  UserCollectedProperties user_collected_properties alignas(16);
   UserCollectedProperties readable_properties;
 
   // convert this object to a human readable form
```

Under ASAN the test then fails, I assume because ASAN uses nonzero padding bytes and we then interpreted it as a string of some nonzero size.

```
$ ./db_sst_test --gtest_filter=DBSSTTest.OpenDBWithInfiniteMaxOpenFilesSubjectToMemoryLimit
...
db_sst_test: cache/cache_reservation_manager.cc:105: rocksdb::Status rocksdb::CacheReservationManagerImpl<R>::ReleaseCacheReservation(std::size_t) [with rocksdb::CacheEntryRole R = rocksdb::CacheEntryRole::kBlockBasedTableReader; std::size_t = long unsigned int]: Assertion `GetTotalMemoryUsed() >= incremental_memory_used' failed.
Received signal 6 (Aborted)
```

Reply:
Let me look into that and see if we should pursue another way of walking through TableProperties

Reply:
The alternative it to iterate up to and including the last `std::string*` given by the (current) last std::string field. That should be safe as long as all the fields in between are std::string. The only downside is that adding another string field at the end of the list will require updating the code.

Reply:
Update: for the purpose of approximating memory, I decided to explicitly enumerating std::string fields of TableProperties for readability as both iteration hacks and explicit enumeration suffer from inflexibility in adding new string prop.

For the purpose of existing TEST_SetRandomTableProperties, since it already exists and requires mutation of field, I opened another PR to fix the issue caused by potential non-zero padding https://github.com/facebook/rocksdb/pull/9812

Will merge this PR after CI is clear. Thanks!